[ADT] Fix llvm::concat_iterator for `ValueT == common_base_class *`

jalopezg-git opened the pull request:
Fix `llvm::concat_iterator` for the case of `ValueT` being a pointer to a common base class to which the result of dereferencing any iterator in `ItersT` can be casted to.  In particular, the case below was not working before this patch, but I see no particular reason why it shouldn't be supported.
```c++
namespace some_namespace {
struct some_struct {
  std::vector<int> data;
  std::string swap_val;
};

struct derives_from_some_struct : some_struct {
};
} // namespace some_namespace

TEST(STLExtrasTest, ConcatRangePtrToDerivedClass) {
  some_namespace::some_struct S0{};
  some_namespace::derives_from_some_struct S1{};
  SmallVector<some_namespace::some_struct *> V0{&S0};
  SmallVector<some_namespace::derives_from_some_struct *> V1{&S1, &S1};

  // Use concat over ranges of pointers to different (but related) types.
  EXPECT_THAT(concat<some_namespace::some_struct *>(V0, V1),
              ElementsAre(&S0, static_cast<some_namespace::some_struct *>(&S1),
                          static_cast<some_namespace::some_struct *>(&S1)));
}
```

Per https://en.cppreference.com/w/cpp/language/implicit_conversion.html, Sec. Pointer conversions,
> A prvalue ptr of type “pointer to (possibly cv-qualified) Derived” can be converted to a prvalue of type “pointer to (possibly cv-qualified) Base”, where Base is a [base class](https://en.cppreference.com/w/cpp/language/derived_class.html) of Derived, and Derived is a [complete](https://en.cppreference.com/w/cpp/language/type-id.html#Incomplete_types) class type.

I.e. conversion shall be possible, but the former implementation of `llvm::concat_iterator` forbids that, given that `handle_type` is pointer-to-pointer type, which is not convertible.

kuhar commented on llvm/unittests/ADT/STLExtrasTest.cpp:
```diff
@@ -532,6 +534,21 @@ TEST(STLExtrasTest, ConcatRangeADL) {
 }
 
+TEST(STLExtrasTest, ConcatRangePtrToDerivedClass) {
+  auto S0 = std::make_unique<some_namespace::some_struct>();
+  auto S1 = std::make_unique<some_namespace::derives_from_some_struct>();
```
I don't think we necessarily need unique_ptr to test this, stack allocated values would work too.

jalopezg-git replied:
Fixed :+1:; thanks for pointing out!

kuhar commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -1088,7 +1089,7 @@ class concat_iterator
       return {};
 
-    if constexpr (ReturnsByValue)
+    if constexpr (ReturnsValueOrPointer)
       return *Begin;
```
So this wraps pointers with `std::optional`? Why can't we return pointers here?

jalopezg-git replied:
Fixed :+1:.

jalopezg-git commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -1032,13 +1032,17 @@ class concat_iterator
+      std::is_pointer_v<ValueT> &&
+      (std::is_convertible_v<decltype(*std::declval<IterTs>()), ValueT> && ...);
 
   using reference_type =
-      typename std::conditional_t<ReturnsByValue, ValueT, ValueT &>;
+      typename std::conditional_t<ReturnsByValue || ReturnsConvertiblePointer,
+                                  ValueT, ValueT &>;
 
-  using handle_type =
-      typename std::conditional_t<ReturnsByValue, std::optional<ValueT>,
-                                  ValueT *>;
+  using handle_type = typename std::conditional_t<
+      ReturnsConvertiblePointer, ValueT,
+      std::conditional_t<ReturnsByValue, std::optional<ValueT>, ValueT *>>;
```
On second thoughts, I cannot see an obvious way to simplify this conditional type; unfortunately, defining a type trait struct would span over more lines (and I don't think it pays off given that this type is only used by a couple of private member functions).

I am not sure it is much clearer, but if it helps, it can be trivially split as
```c++
  using wrap_type =
      typename std::conditional_t<ReturnsByValue, std::optional<ValueT>, ValueT *>;
  using handle_type = typename std::conditional_t<ReturnsConvertiblePointer, ValueT, wrap_type>;
```
Any preference?  Other than that, I think the PR is ready; build passing :+1:.

jalopezg-git replied:
I have split that into the types below and added a clarifying comment.  PTAL.
```c++
  using optional_value_type =
      std::conditional_t<ReturnsByValue, std::optional<ValueT>, ValueT *>;
  using handle_type = typename std::conditional_t<ReturnsConvertiblePointer,
                                                  ValueT, optional_value_type>;
```

jalopezg-git replied:
Resolving conversation as N/A due to: `handle_type` has been removed in commit https://github.com/llvm/llvm-project/pull/144744/commits/cd1127c3f7d7b3f7a2d30cb7fa55f4249dcde803.

kuhar commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -1032,13 +1032,17 @@ class concat_iterator
   static constexpr bool ReturnsByValue =
       !(std::is_reference_v<decltype(*std::declval<IterTs>())> && ...);
+  static constexpr bool ReturnsConvertiblePointer =
+      std::is_pointer_v<ValueT> &&
+      (std::is_convertible_v<decltype(*std::declval<IterTs>()), ValueT> && ...);
 
   using reference_type =
-      typename std::conditional_t<ReturnsByValue, ValueT, ValueT &>;
+      typename std::conditional_t<ReturnsByValue || ReturnsConvertiblePointer,
+                                  ValueT, ValueT &>;
```
Doesn't this change the reference type in the case where all pointer types are the same? Prior to this PR, the `reference_type` used to be `T *&`, now it will become `T *`.

kuhar replied:
Similar for `handle_type`, doesn't this change the type from `T **` to `T *`?

dwblaikie replied:
some test coverage would be good either way - to demonstrate what breaks if such changes were to occur (ie: assignability of a dereference expression (`*concat_iter = other_ptr`) would fail to compile if the reference type changed from `T*&` to `T*`)

jalopezg-git replied:
> some test coverage would be good either way - to demonstrate what breaks if such changes were to occur (ie: assignability of a dereference expression (`*concat_iter = other_ptr`) would fail to compile if the reference type changed from `T*&` to `T*`)

Not sure I completely understood your intent; however, I have added one additional test that checks that returning by reference where appropriate is still okay.
Such pattern can also been seen, e.g., in https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp#L217.

PTAL and let me know if I still need to apply some changes to this PR.

kuhar reviewed (commented):
I took a closer look and I think I understand what the underlying issue was: with mixed pointer types, reference/handle type can no longer be `T*&` / `T**` because while it's fine to cast derived pointers to `T*`, we can't take an address off them using that base type.

I wonder if we should tighten this check to make sure that at least one input range type doesn't match the `ValueT` pointer type, and with that having them be returned via the same logic as `ReturnsByValue` doesn't sounds too bad to me. WDYT?

kuhar commented:
I think the two pieces of feedback haven't been addressed yet:
1. I think we should test whether the values we iterate over are references to elements or not
2. For containers of the same pointer type as in `concat<T *>`, I think we should keep existing behavior and allow for modification

jalopezg-git commented:
Apologies for coming back to this w/ almost a 3-month delay :sweat_smile:; it has been quite hectic due to work in a different project.

@kuhar `handle_type` was a major cause of headaches here; therefore, I first simplified `lllvm::concat_iterator` to get rid of it.  IMHO, the result is simpler / more maintainable. See commit https://github.com/llvm/llvm-project/pull/144744/commits/cd1127c3f7d7b3f7a2d30cb7fa55f4249dcde803.

> I think the two pieces of feedback in [#144744 (comment)](https://github.com/llvm/llvm-project/pull/144744#discussion_r2159675066) haven't been addressed yet:
> 
>     1. I think we should test whether the values we iterate over are references to elements or not
> 

Tests have been updated; in particular, PTAL at https://github.com/llvm/llvm-project/pull/144744/files#diff-3ec26250dd1ddfd2682b4abed65da0d8cbb509991fc1a7a3795bb29389e3da6cR548 and https://github.com/llvm/llvm-project/pull/144744/files#diff-3ec26250dd1ddfd2682b4abed65da0d8cbb509991fc1a7a3795bb29389e3da6cR569.

>     2. For containers of the same pointer type as in `concat<T *>`, I think we should keep existing behavior and allow for modification

In case all `IterTs...` yield the same value type as `ValueT` (see https://github.com/llvm/llvm-project/pull/144744/files#diff-d6563ab6a854637e8bc9e97fe718cd19dda63da0321091e9f0bf71ecb20d8003R999), `reference_type` is a reference type and allows for container modification.
See test at https://github.com/llvm/llvm-project/pull/144744/files#diff-3ec26250dd1ddfd2682b4abed65da0d8cbb509991fc1a7a3795bb29389e3da6cR550.

jalopezg-git commented:
@kuhar, thank you for your previous feedback -- much appreciated!

@CarlosAlbertoEnciso , @jmorse , @kuhar This pull request is now ready for another round of review :+1:.

kuhar commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -996,13 +996,18 @@ class concat_iterator
       !(std::is_reference_v<decltype(*std::declval<IterTs>())> && ...);
-
+  static constexpr bool ReturnsConvertibleType =
+      (!std::is_same_v<
+           std::remove_reference_t<decltype(*std::declval<IterTs>())>,
+           ValueT> ||
```
This is quite complicated and it's not clear to me what the negation operator applies to. Maybe hoist this to a helper struct and add tests with static_assert?

kuhar commented on llvm/unittests/ADT/STLExtrasTest.cpp:
```diff
@@ -532,6 +534,44 @@ TEST(STLExtrasTest, ConcatRangeADL) {
+      std::is_same_v<decltype(*C.begin()), some_namespace::some_struct *&>);
+  EXPECT_THAT(C, ElementsAre(&S0, &S1, &S1));
+  // `reference_type` should still allow container modification.
+  llvm::for_each(C, [&S0](auto &i) {
+    if (i == &S0)
+      i = nullptr;
+  });
```
Could we use a regular range for loop?

jalopezg-git replied:
It was meant to be a one-liner, but `clang-format` broke lines this way.  Fixed to use range-`for` :+1:.

jalopezg-git commented on llvm/unittests/ADT/STLExtrasTest.cpp:
```diff
@@ -532,6 +534,44 @@ TEST(STLExtrasTest, ConcatRangeADL) {
+
+  // Dereferencing all iterators yields different (but convertible types);
+  // conversion takes place, `reference_type` is `some_namespace::some_struct
+  // *`.
```
`clang-format`'s decision :sweat_smile: -- fixed.

kuhar commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -978,6 +978,17 @@ detail::zip_longest_range<T, U, Args...> zip_longest(T &&t, U &&u,
       std::forward<T>(t), std::forward<U>(u), std::forward<Args>(args)...);
 }
 
+/// `all_types_same_as::value == true` if all types in `OtherTs` are same as
+/// `FirstT`, ignoring cv-qualifiers.
+template <typename FirstT, typename... OtherTs> struct all_types_same_as {
+  static constexpr bool value =
+      (std::is_same_v<std::remove_cv_t<OtherTs>, std::remove_cv_t<FirstT>> &&
+       ...);
+};
+template <typename FirstT, typename... OtherTs>
+constexpr bool all_types_same_as_v =
+    all_types_same_as<FirstT, OtherTs...>::value;
```
Alternatively, you could also have `all_type_match_with<predicate, type0, type1, ...>` and then define `using all_types_equal = all_type_match_with<std::is_same, type0, type1, ...>;`

kuhar replied:
Also, this can be defined ass true for zero types

jalopezg-git replied:
Resolving conversation (as the item has been addressed in the latest revision).

kuhar reviewed (commented):
Looks good, just a couple of minor issues

jalopezg-git committed:
[ADT] Fix llvm::concat_iterator for `ValueT == common_base_class *`

Fix llvm::concat_iterator for the case of `ValueT` being a pointer
to a common base class to which the result of dereferencing any
iterator in `ItersT` can be casted to.

jalopezg-git committed:
[ADT] Simplify `llvm::concat_iterator` implementation

Simplify `llvm::concat_iterator::increment()` and `llvm::concat_iterator::get()`.
This makes it possible to also get rid of `handle_type`, which was causing
some additional complications.

jalopezg-git commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -996,13 +1003,17 @@ class concat_iterator
-
+  static constexpr bool ReturnsConvertibleType =
+      !all_types_equal_v<std::remove_cv_t<ValueT>,
+                         std::remove_cv_t<std::remove_reference_t<
```
Thank you :+1: -- I was not aware of `remove_cvref_t`!

jalopezg-git commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -996,13 +1003,17 @@ class concat_iterator
-  using handle_type =
-      typename std::conditional_t<ReturnsByValue, std::optional<ValueT>,
-                                  ValueT *>;
+      typename std::conditional_t<ReturnsByValue || ReturnsConvertibleType,
```
It should not be needed; removed.  PTAL and feel free to approve it it is all good :+1:.

jalopezg-git commented:
Thank you for the invaluable feedback, @kuhar!
